xbmc/addons/kodi-dev-kit/include/kodi/tools/EndTime.h

```diff
@@ -0,0 +1,198 @@
+///     }
+///   }
+///
+/// };
+/// ~~~~~~~~~~~~~
+///
+///@{
+class CEndTime
+{
+public:
+  //============================================================================
+  /// @ingroup cpp_kodi_tools_CEndTime
+  /// @brief Class constructor with no time to expiry set
+  ///
+  inline CEndTime() = default;
+  //============================================================================
+  /// @ingroup cpp_kodi_tools_CEndTime
+  /// @brief Class constructor to set future time when timer has expired
+  ///
+  /// @param[in] millisecondsIntoTheFuture the time in the future we cosider this timer as expired
+  ///
+  inline explicit CEndTime(unsigned int millisecondsIntoTheFuture)
+    : m_startTime(std::chrono::system_clock::now().time_since_epoch()),
+      m_totalWaitTime(std::chrono::milliseconds(millisecondsIntoTheFuture))
+  {
+  }
+
+  //============================================================================
+  /// @ingroup cpp_kodi_tools_CEndTime
+  /// @brief Set the time in the future we cosider this timer as expired
+  ///
+  /// @param[in] millisecondsIntoTheFuture the time in the future we cosider this timer as expired
+  ///
+  inline void Set(unsigned int millisecondsIntoTheFuture)
+  {
+    using namespace std::chrono;
+
+    m_startTime = system_clock::now().time_since_epoch();
+    m_totalWaitTime = milliseconds(millisecondsIntoTheFuture);
+  }
+
+  //============================================================================
+  /// @ingroup cpp_kodi_tools_CEndTime
+  /// @brief Check if the expiry time has been reached
+  ///
+  /// @return True if the expiry amount of time has past, false otherwise
+  ///
+  inline bool IsTimePast() const
+  {
+    using namespace std::chrono;
+
+    return m_totalWaitTime.count() == std::numeric_limits<unsigned int>::max()
+               ? false
+               : (m_totalWaitTime.count() == 0
+                      ? true
+                      : (system_clock::now().time_since_epoch() - m_startTime) >= m_totalWaitTime);
+  }
+
+  //============================================================================
+  /// @ingroup cpp_kodi_tools_CEndTime
+  /// @brief The amount of time left till this timer expires
+  ///
+  /// @return 0 if the expiry amount of time has past, the numbe rof milliseconds remaining otherwise
+  ///
+  inline unsigned int MillisLeft() const
+  {
+    using namespace std::chrono;
+
+    if (m_totalWaitTime.count() == std::numeric_limits<unsigned int>::max())
+      return std::numeric_limits<unsigned int>::max();
+    if (m_totalWaitTime.count() == 0)
+      return 0;
+    unsigned int timeWaitedAlready =
+        duration_cast<milliseconds>(std::chrono::system_clock::now().time_since_epoch() -
+                                    m_startTime)
+            .count();
+    return (timeWaitedAlready >= m_totalWaitTime.count())
+               ? 0
+               : (m_totalWaitTime.count() - timeWaitedAlready);
```

Comment:
This is an instance where _machine readability_ (clang-format's effort to apply line length limits to improve readability and accessibility) actually hinders _human readability_.

This section would be much more human-readable with simply using long lines. However, long lines are hard to read on screens with large text, hurting accessibility.

Some refactoring can achieve both goals simultaneously. How about:

```c++
if (m_totalWaitTime.count() == std::numeric_limits<unsigned int>::max())
  return std::numeric_limits<unsigned int>::max();

if (m_totalWaitTime.count() == 0)
  return 0;

auto elapsed = system_clock::now().time_since_epoch() - m_startTime;
unsigned int timeWaitedAlready = duration_cast<milliseconds>(elapsed).count();

if (timeWaitedAlready >= m_totalWaitTime.count())
  return 0;

return m_totalWaitTime.count() - timeWaitedAlready;
```

(please test, I've been known to lead people astray with untested comment code :wink: )

Reply:
Great, suggestion. Thanks!